diff -ru ./board/raspberrypi3-64.orig/genimage-raspberrypi3-64.cfg ./board/raspberrypi3-64/genimage-raspberrypi3-64.cfg
--- ./board/raspberrypi3-64.orig/genimage-raspberrypi3-64.cfg	2017-06-06 20:17:53.272784758 +0200
+++ ./board/raspberrypi3-64/genimage-raspberrypi3-64.cfg	2017-06-06 20:15:11.272592858 +0200
@@ -1,13 +1,15 @@
 image boot.vfat {
   vfat {
     files = {
-      "bcm2710-rpi-3-b.dtb",
+#      "bcm2710-rpi-3-b.dtb",
       "bcm2837-rpi-3-b.dtb",
       "rpi-firmware/bootcode.bin",
       "rpi-firmware/cmdline.txt",
       "rpi-firmware/config.txt",
       "rpi-firmware/fixup.dat",
       "rpi-firmware/start.elf",
+      "u-boot.bin",
+      "boot.scr",
       "Image"
     }
   }
Seulement dans ./board/raspberrypi3-64: raspberrypi
Seulement dans ./board/raspberrypi3-64: raspberrypi3
