diff -Nru board/raspberrypi3-64.orig/genimage-raspberrypi3-64.cfg board/raspberrypi3-64/genimage-raspberrypi3-64.cfg
--- board/raspberrypi3-64.orig/genimage-raspberrypi3-64.cfg	2017-06-05 19:21:16.221272013 +0200
+++ board/raspberrypi3-64/genimage-raspberrypi3-64.cfg	2017-06-05 19:20:07.157643317 +0200
@@ -1,7 +1,7 @@
 image boot.vfat {
   vfat {
     files = {
-      "bcm2710-rpi-3-b.dtb",
+#      "bcm2710-rpi-3-b.dtb",
       "bcm2837-rpi-3-b.dtb",
       "rpi-firmware/bootcode.bin",
       "rpi-firmware/cmdline.txt",
