--- board/raspberrypi3.orig/genimage-raspberrypi3.cfg	2017-06-09 22:27:58.591225031 +0200
+++ board/raspberrypi3/genimage-raspberrypi3.cfg	2017-06-09 22:29:47.938870874 +0200
@@ -1,8 +1,8 @@
 image boot.vfat {
   vfat {
     files = {
-      "bcm2710-rpi-3-b.dtb",
-      "bcm2710-rpi-cm3.dtb",
+      "bcm2837-rpi-3-b.dtb",
+# "bcm2710-rpi-cm3.dtb",
       "rpi-firmware/bootcode.bin",
       "rpi-firmware/cmdline.txt",
       "rpi-firmware/config.txt",
